storage/innobase/trx/trx0trx.cc

```diff
@@ -623,11 +624,52 @@ trx_resurrect_table_locks(
 		trx_undo_rec_get_pars(
 			undo_rec, &type, &cmpl_info,
 			&updated_extern, &undo_no, &table_id);
-		tables.insert(table_id);
-
-		undo_rec = trx_undo_get_prev_rec(
-			undo_rec, undo->hdr_page_no,
-			undo->hdr_offset, false, &mtr);
+              /**
+               * If we are doing a restore, export, then it is not guaranteed that all the tables
+               * will be present in the backup. However, we also need to rollback uncommitted
+               * transactions from the existing tables. Therefore, we collect only the tables
+               * that exists in the data directory when going through undo log records.
+               * Undo log records are applied only to the exported tables.
+               * See MDEV-29050 for more details.
+               */
+              // is this restore export ? -- then, add table only if it exists in the data directory ...
+              if ((srv_operation == SRV_OPERATION_RESTORE_EXPORT) &&
+                      srv_file_per_table) {
+
+                     char *table_name = NULL;
+
+                     // find the table name for the given table id ...
+                     mutex_enter(&dict_sys.mutex);
+                     dict_find_table_name_from_id(table_id, &table_name);
+                     mutex_exit(&dict_sys.mutex);
+
+                     // if table name is not null and if the table is exported to the backup
+                     // then add it to the table list ...
+                     if (table_name != NULL) {
+                            // given table name, find the ibd file path ...
+                            char *ibd_file_path = fil_make_filepath(NULL, table_name, IBD, false);
```

Comment:
Even if the data file was excluded from a backup, the table metadata will be available in `SYS_TABLES` and other dictionary tables.

This code seems to be duplicating some existing logic. I do not think that this is a good idea. It would be better to tweak `dict_load_table_one()` or `dict_load_table_low()` or related code so that it does the right thing.